### PR TITLE
Hierarchical Find References

### PR DESCRIPTION
**Pull request by dpoeschl**
In this initial implementation, find references results are now grouped by the definition that they directly reference.

For symbols defined in metadata, we previously showed the references without showing the definition itself. We now show the metadata definition as well (but grayed out).

Two linked file adjustments:
- all definitions show their containing project, so linked definitions can now be distinguished (see the two `Class1.ToString()` definitions below)
- linked references *to the same symbol* show the project name to disambiguate. (see the first two references to `int.ToString()` below)

Here's a screenshot:
image

And for reference, here's the "before" picture from the same symbol:
image

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/Library/FindResults/LibraryManager_FindReferences.cs [hunk 1]
```diff
@@ -50,30 +51,59 @@ private IList<AbstractListItem> CreateFindReferencesItems(Solution solution, IEn
...
             HashSet<ValueTuple<Document, TextSpan>> uniqueLocations,
-            IEnumerable<Location> locations,
+            Location location,
             Glyph glyph)
         {
+            if (!location.IsInSource)
+            {
+                return referencedSymbol.Locations.Any()
+                    ? new MetadataDefinitionTreeItem(referencedSymbol.Definition, glyph.GetGlyphIndex())
+                    : null;
```
Comment (jaredpar) [inline src/VisualStudio/Core/Def/Implementation/Library/FindResults/LibraryManager_FindReferences.cs]: Intuitively I would never expect a method named `CreatedX` to return `null`.

**Comment by sharwell** [comment]
For reference, can you include a screenshot or diagram to show the new organization? :smile:

File: src/VisualStudio/Core/Def/Implementation/Library/FindResults/TreeItems/ExternalTreeItem.cs [hunk 2]
```diff
@@ -9,14 +9,14 @@
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.Library.FindResults
 {
-    internal class ExternalListItem : AbstractListItem
+    internal class ExternalTreeItem : AbstractTreeItem
```
Comment (jaredpar) [inline src/VisualStudio/Core/Def/Implementation/Library/FindResults/TreeItems/ExternalTreeItem.cs]: What does `External` mean here?

File: src/VisualStudio/Core/Def/Implementation/Library/FindResults/LibraryManager_FindReferences.cs (hunk 1 above)
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/Library/FindResults/LibraryManager_FindReferences.cs]: :+1: Maybe ConvertToDefinitionItem?

File: src/VisualStudio/Core/Def/Implementation/Library/FindResults/TreeItems/ExternalTreeItem.cs (hunk 2 above)
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/Library/FindResults/TreeItems/ExternalTreeItem.cs]: From the use above, it means "From outside Roslyn", as in, when the Xaml language service gives us definitions in Xaml files, etc.

**Comment by Pilchie** [comment]
:+1:

File: src/VisualStudio/Core/Def/Implementation/Library/FindResults/TreeItems/ExternalTreeItem.cs (hunk 2 above)
Comment (jaredpar) [inline src/VisualStudio/Core/Def/Implementation/Library/FindResults/TreeItems/ExternalTreeItem.cs]: Can we add a comment to that effect here?

Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/Library/FindResults/TreeItems/ExternalTreeItem.cs]: Yes, will comment. I'm also considering renaming it to `ExternalLanguageDefinitionTreeItem`

**Comment by dpoeschl** [comment]
@sharwell Screenshot added to top comment. There's more we'd like to do, but it might involve some work on the Find Symbol Results window itself.

**Comment by Pilchie** [comment]
@dpoeschl Just had a thought - why doesn't navigating to a metadata definition show metadata as source? Could certainly be added later of course.

**Comment by dpoeschl** [comment]
@Pilchie Yes, we can consider doing that later. It's not yet implemented because I cannot hold on to the symbol because these items can persist indefinitely, and the main Metadata as Source entrypoint is symbol-based.

**Comment by DustinCampbell** [comment]
SymbolId?

File: src/VisualStudio/Core/Def/Implementation/Library/FindResults/LibraryManager_FindReferences.cs (hunk 1 above)
Comment (DustinCampbell) [inline src/VisualStudio/Core/Def/Implementation/Library/FindResults/LibraryManager_FindReferences.cs]: :+1:

**Comment by dpoeschl** [comment]
@DustinCampbell I assume you mean [SymbolKey](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/SymbolId/SymbolKey.cs,91ee6a4b910afd68), and yeah, it'll have to be done that way.

**Comment by DustinCampbell** [comment]
Yeah [SymbolKey](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/SymbolId/SymbolKey.cs,91ee6a4b910afd68). I forgot we changed the name.

**Comment by DustinCampbell** [comment]
:+1:

**Comment by DustinCampbell** [comment]
Looking forward to seeing some reduction of those long file paths though. :smile:

**Comment by dpoeschl** [comment]
@DustinCampbell Yeah, we'll need to think about how best to do that depending on whether I can further group references by file or not.

**Comment by DustinCampbell** [comment]
It's pretty clear in my head. Are there edge cases you're concerned about?

File: src/VisualStudio/Core/Def/Implementation/Library/FindResults/LibraryManager_FindReferences.cs [hunk 3]
```diff
@@ -50,30 +51,59 @@ private IList<AbstractListItem> CreateFindReferencesItems(Solution solution, IEn
...
+            ReferencedSymbol referencedSymbol,
             HashSet<ValueTuple<Document, TextSpan>> uniqueLocations,
-            IEnumerable<Location> locations,
+            Location location,
             Glyph glyph)
         {
+            if (!location.IsInSource)
+            {
+                return referencedSymbol.Locations.Any()
+                    ? new MetadataDefinitionTreeItem(referencedSymbol.Definition, glyph.GetGlyphIndex())
```
Comment (balajikris) [inline src/VisualStudio/Core/Def/Implementation/Library/FindResults/LibraryManager_FindReferences.cs]: Do we get references to hidden locations as well or are they already filtered out?

**Comment by balajikris** [comment]
Looks good. Test coverage for the new types and functions that were added?

**Comment by sharwell** [comment]
This is a really cool feature for `virtual` elements. :+1:

**Pull request merged**